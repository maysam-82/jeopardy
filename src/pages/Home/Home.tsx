import React from 'react';
import { connect } from 'react-redux';
import { Spinner } from 'react-bootstrap';
import { ICategory } from '../../types/category.d';
import { IStoreState } from '../../redux/reducers/index';
import { getCategories, setCategory } from '../../redux/actions/categories';
import history from '../../history';
import Categories from '../../components/Categories/Categories';

export interface IHomeProps {
	getCategories: Function;
	categories: ICategory[];
	setCategory: typeof setCategory;
	isFetching: boolean;
}

export class Home extends React.Component<IHomeProps, {}> {
	componentDidMount() {
		const { categories, getCategories } = this.props;
		if (categories.length === 0) {
			getCategories();
		}
	}
	categoryClickHandler = (categoryId: number) => {
		const { setCategory, categories } = this.props;
		const category = categories.find((category) => category.id === categoryId);
		if (category) {
			setCategory(category);
			history.push('/category');
		}
	};

	render() {
		const { isFetching, categories } = this.props;
		return (
			<div className="mb-2">
				{isFetching ? (
					<div className="spinner">
						<Spinner animation="border" role="status" variant="secondary" />
					</div>
				) : (
					<Categories
						categories={categories}
						categoryClick={this.categoryClickHandler}
					/>
				)}
			</div>
		);
	}
}

const mapStateToProps = (state: IStoreState) => ({
	categories: state.category.categories,
	isFetching: state.category.isFetching,
});
export default connect(mapStateToProps, { getCategories, setCategory })(Home);
