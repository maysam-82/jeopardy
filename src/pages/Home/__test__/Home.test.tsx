import React from 'react';
import { shallow, ShallowWrapper, mount, ReactWrapper } from 'enzyme';
import { Home, IHomeProps } from '../Home';
import { testCategories } from '../../../fixtures/testData';

const testHomeProps: IHomeProps = {
	getCategories: jest.fn(),
	categories: testCategories,
	setCategory: jest.fn(),
	isFetching: false,
};

describe('Home page component', () => {
	let component: ReactWrapper<IHomeProps, {}, Home>;
	beforeEach(() => {
		component = mount<Home, IHomeProps, {}>(<Home {...testHomeProps} />);
	});
	it('should render Home page', () => {
		expect(component).toMatchSnapshot();
	});
	it('should not call `getCategories` function', () => {
		expect(component.instance().props.getCategories).toHaveBeenCalledTimes(0);
	});
	describe('When clicking on category', () => {
		it('should call `setCategory` with selected category data', () => {
			component.find('ListGroupItem').at(0).simulate('click');
			expect(component.instance().props.setCategory).toHaveBeenCalledWith(
				testHomeProps.categories[0]
			);
		});
	});
	describe('When there is no category in `categories`', () => {
		let componentWithNoCategory: ShallowWrapper<IHomeProps, {}, Home>;
		const propsWithNoCategory = { ...testHomeProps, categories: [] };
		beforeEach(() => {
			componentWithNoCategory = shallow<Home, IHomeProps, {}>(
				<Home {...propsWithNoCategory} />
			);
		});

		it('should call `getCategories` function', () => {
			expect(
				componentWithNoCategory.instance().props.getCategories
			).toHaveBeenCalledTimes(1);
		});
	});
});
