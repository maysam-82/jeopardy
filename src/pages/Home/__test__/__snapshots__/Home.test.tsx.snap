// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page component should render Home page 1`] = `
<Home
  categories={
    Array [
      Object {
        "clues_count": 5,
        "id": 11531,
        "title": "test category 001",
      },
      Object {
        "clues_count": 5,
        "id": 11532,
        "title": "test category 002",
      },
      Object {
        "clues_count": 10,
        "id": 5412,
        "title": "test category 003",
      },
    ]
  }
  getCategories={[MockFunction]}
  isFetching={false}
  setCategory={[MockFunction]}
>
  <div
    className="mb-2"
  >
    <Categories
      categories={
        Array [
          Object {
            "clues_count": 5,
            "id": 11531,
            "title": "test category 001",
          },
          Object {
            "clues_count": 5,
            "id": 11532,
            "title": "test category 002",
          },
          Object {
            "clues_count": 10,
            "id": 5412,
            "title": "test category 003",
          },
        ]
      }
      categoryClick={[Function]}
    >
      <ListGroup
        horizontal={null}
        variant={null}
      >
        <ForwardRef
          as="div"
          className="list-group"
          onSelect={[Function]}
        >
          <div
            className="list-group"
            onKeyDown={[Function]}
          >
            <ListGroupItem
              action={true}
              active={false}
              disabled={false}
              key="11531"
              onClick={[Function]}
              variant="secondary"
            >
              <ForwardRef
                as="button"
                className="list-group-item list-group-item-secondary list-group-item-action"
                disabled={false}
                eventKey={null}
                onClick={[Function]}
              >
                <button
                  className="list-group-item list-group-item-secondary list-group-item-action"
                  data-rb-event-key={null}
                  disabled={false}
                  onClick={[Function]}
                >
                  test category 001
                </button>
              </ForwardRef>
            </ListGroupItem>
            <ListGroupItem
              action={true}
              active={false}
              disabled={false}
              key="11532"
              onClick={[Function]}
              variant="secondary"
            >
              <ForwardRef
                as="button"
                className="list-group-item list-group-item-secondary list-group-item-action"
                disabled={false}
                eventKey={null}
                onClick={[Function]}
              >
                <button
                  className="list-group-item list-group-item-secondary list-group-item-action"
                  data-rb-event-key={null}
                  disabled={false}
                  onClick={[Function]}
                >
                  test category 002
                </button>
              </ForwardRef>
            </ListGroupItem>
            <ListGroupItem
              action={true}
              active={false}
              disabled={false}
              key="5412"
              onClick={[Function]}
              variant="secondary"
            >
              <ForwardRef
                as="button"
                className="list-group-item list-group-item-secondary list-group-item-action"
                disabled={false}
                eventKey={null}
                onClick={[Function]}
              >
                <button
                  className="list-group-item list-group-item-secondary list-group-item-action"
                  data-rb-event-key={null}
                  disabled={false}
                  onClick={[Function]}
                >
                  test category 003
                </button>
              </ForwardRef>
            </ListGroupItem>
          </div>
        </ForwardRef>
      </ListGroup>
    </Categories>
  </div>
</Home>
`;
